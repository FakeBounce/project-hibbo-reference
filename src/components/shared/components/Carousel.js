import React, { Component } from 'react';
import { Dimensions, View, ScrollView } from 'react-native';
import { arrayOf, func, object } from 'prop-types';
import { bind } from 'decko';
import EStyleSheet from 'react-native-extended-stylesheet';

const styles = EStyleSheet.create({
  wrapper: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  scrollView: {
    marginBottom: 30,
  },
  thumbnailsWrapper: {
    flexDirection: 'row',
  },
});

class Carousel extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentItem: 0,
    };
  }

  @bind
  onScroll(event) {
    this.setState({
      currentItem: Math.round(
        event.nativeEvent.contentOffset.x / Dimensions.get('window').width,
      ),
    });
  }

  @bind
  renderThumbnail(item, index) {
    if (this.state.currentItem === index) {
      return this.props.renderActiveThumbnail(item, index);
    }

    return this.props.renderEmptyThumbnail(item, index);
  }

  render() {
    return (
      <View style={styles.wrapper}>
        <ScrollView
          horizontal
          showsHorizontalScrollIndicator={false}
          snapToInterval={Dimensions.get('window').width}
          // @todo: Doesn't work on android, see pairing to correct
          snapToAlignment="start"
          decelerationRate={0}
          scrollEventThrottle={3}
          onScroll={this.onScroll}
          style={styles.scrollView}
        >
          {this.props.items.map(this.props.renderItem)}
        </ScrollView>

        <View style={styles.thumbnailsWrapper}>
          {this.props.items.map(this.renderThumbnail)}
        </View>
      </View>
    );
  }
}

Carousel.propTypes = {
  items: arrayOf(object).isRequired,
  renderItem: func.isRequired,
  renderActiveThumbnail: func.isRequired,
  renderEmptyThumbnail: func.isRequired,
};

export default Carousel;
