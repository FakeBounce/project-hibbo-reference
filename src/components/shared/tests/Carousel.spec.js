import { Dimensions, View, Image } from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import EStyleSheet from 'react-native-extended-stylesheet';

import bigPiggy from 'assets/settings/big-piggy.png';
import Carousel from '../components/Carousel';

const { width } = Dimensions.get('window');

const styles = EStyleSheet.create({
  slideWrapper: {
    alignItems: 'center',
  },
  thumbnailItemWrapper: {
    borderRadius: 16,
    borderColor: '$colors.lightGrey',
    padding: 2,
    margin: 5,
  },
  activeThumbnailItemWrapper: {
    borderWidth: 1,
  },
  thumbnail: {
    width: 16,
    height: 16,
    borderRadius: 8,
    backgroundColor: '$colors.lightGrey',
  },
});

const carouselItems = [{ uri: bigPiggy }, { uri: bigPiggy }, { uri: bigPiggy }];

const renderItem = (item, index) => (
  <View key={index} style={[styles.slideWrapper, { width }]}>
    <Image source={item.uri} />
  </View>
);

const renderActiveThumbnail = (item, index) => (
  <View
    key={index}
    style={[styles.thumbnailItemWrapper, styles.activeThumbnailItemWrapper]}
  >
    <View style={styles.thumbnail} />
  </View>
);

const renderEmptyThumbnail = (item, index) => (
  <View key={index} style={styles.thumbnailItemWrapper}>
    <View style={styles.thumbnail} />
  </View>
);

describe('Carousel', () => {
  it('should render the component', () => {
    const carousel = renderer
      .create(
        <Carousel
          items={carouselItems}
          renderItem={renderItem}
          renderActiveThumbnail={renderActiveThumbnail}
          renderEmptyThumbnail={renderEmptyThumbnail}
        />,
      )
      .toJSON();

    expect(carousel).toMatchSnapshot();
  });
});
