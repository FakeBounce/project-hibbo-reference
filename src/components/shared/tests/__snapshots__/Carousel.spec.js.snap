// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render the component 1`] = `
<View
  style={undefined}
>
  <RCTScrollView
    decelerationRate={0}
    horizontal={true}
    onScroll={[Function]}
    scrollEventThrottle={3}
    showsHorizontalScrollIndicator={false}
    snapToAlignment="start"
    snapToInterval={750}
    style={undefined}
  >
    <View>
      <View
        style={
          Array [
            undefined,
            Object {
              "width": 750,
            },
          ]
        }
      >
        <Image
          source={1}
        />
      </View>
      <View
        style={
          Array [
            undefined,
            Object {
              "width": 750,
            },
          ]
        }
      >
        <Image
          source={1}
        />
      </View>
      <View
        style={
          Array [
            undefined,
            Object {
              "width": 750,
            },
          ]
        }
      >
        <Image
          source={1}
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={undefined}
  >
    <View
      style={
        Array [
          undefined,
          undefined,
        ]
      }
    >
      <View
        style={undefined}
      />
    </View>
    <View
      style={undefined}
    >
      <View
        style={undefined}
      />
    </View>
    <View
      style={undefined}
    >
      <View
        style={undefined}
      />
    </View>
  </View>
</View>
`;
